cmd/oci-lifecycle-demo/main.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"fmt"
+
+	l "github.com/opencontainers/runtime-tools/lifecycle"
+)
+
+func TestDemo() {
+	runtime := "runc"
+
+	type ActionCase struct {
+		action   l.LifecycleAction
+		args     []string
+		expected bool
+	}
+	cases := []struct {
+		bundle  string
+		id      string
+		actions []ActionCase
+	}{
+		{"noexist_bundle", "1", []ActionCase{
+			{action: l.LifecycleCreate, expected: false}},
+		},
+		{"exist_bundle", "1", []ActionCase{
+			{action: l.LifecycleCreate, expected: true}},
+		},
+		{"exist_bundle", "1", []ActionCase{
+			{action: l.LifecycleCreate, expected: true},
+			{action: l.LifecycleCreate, expected: false}},
+		},
+		{"exist_bundle_run_err", "1", []ActionCase{
+			{action: l.LifecycleCreate, expected: true},
+			{action: l.LifecycleStart, expected: false}},
```

Comment:
The sharness analog to this is:

```shell
test_expect_success 'create fails when bundle does not exist' "
  test_expect_code 1 ${RUNTIME} create --bundle noexist_bundle 1
"

test_expect_success 'create/kill/destroy succeeds when bundle exists' "
  subreaper-wrapper ${RUNTIME} create --bundle exist_bundle 1 &
  while test ! -e 1-created; do sleep 1; done &&  # this would be easier with runtime-spec#508, but works as a stop-gap with subreaper-wrapper touching this file when 'create' exits
  ${RUNTIME} kill 1 &&
  wait &&
  ${RUNTIME} delete 1 &&
"

test_expect_success 'second create call with same container ID fails' "
  subreaper-wrapper ${RUNTIME} create --bundle exist_bundle 1 &
  while test ! -e 1-created; do sleep 1; done &&  # this would be easier with runtime-spec#508, but works as a stop-gap with subreaper-wrapper touching this file when 'create' exits
  test_expect_code 1 ${RUNTIME} create --bundle exist_bundle 1
  ${RUNTIME} kill 1 &&
  wait &&
  ${RUNTIME} delete 1 &&
"

test_expect_success 'exist_bundle_run_err fails' "
  subreaper-wrapper ${RUNTIME} create --bundle exist_bundle_run_err 1 &
  while test ! -e 1-created; do sleep 1; done &&  # this would be easier with runtime-spec#508, but works as a stop-gap with subreaper-wrapper touching this file when 'create' exits
  test_expect_code 1 wait &&
  ${RUNTIME} delete 1 &&
"
```

That's a bit more complicated than your `cases` literal here, but:

* There's no need to write code to iterate over the cases,
* There's no need for an analog to your `Lifecycle` type,
* You aren't killing successful creates before deleting them, and
* You don't do any of the [subreaper magic][1] that you need if you want to see when/how the container process exits.

And I'm cutting corners by assuming that failing commands will exit 1 (opencontainers/runtime-spec#513 only requires non-zero).  To do this right in sharness, I'd want to define a `test_expect_code_not` helper.

[1]: https://github.com/opencontainers/runtime-tools/pull/232#discussion_r82655833